plugins/hydra_ax_sweeper/hydra_plugins/hydra_ax_sweeper/ax_sweeper.py

```diff
@@ -0,0 +1,303 @@
+        self.early_stopper = EarlyStopper(**early_stop)
+        self.verbose_logging = verbose_logging
+        self.random_seed = random_seed
+        self.max_trials = max_trials
+        self.ax_params = ax_params
+        self.sweep_dir: str
+
+    def setup(
+        self,
+        config: DictConfig,
+        config_loader: ConfigLoader,
+        task_function: TaskFunction,
+    ) -> None:
+        self.launcher = Plugins.instantiate_launcher(
+            config=config, config_loader=config_loader, task_function=task_function
+        )
+        self.sweep_dir = config.hydra.sweep.dir
+
+    def sweep(self, arguments: List[str]) -> None:
+        ax_client = self.setup_ax_client(arguments)
+
+        num_trials_left = self.max_trials
+        recommended_max_parallelism = ax_client.get_recommended_max_parallelism()
+        current_parallelism_index = 0
+        # Index to track the parallelism value we are using right now.
+
+        while num_trials_left > 0:
+            current_parallelism = recommended_max_parallelism[current_parallelism_index]
+            num_trials, max_parallelism_setting = current_parallelism
+            num_trials_so_far = 0
+            while (
+                num_trials > num_trials_so_far or num_trials == -1
+            ) and num_trials_left > 0:
+                batch_of_trials = get_one_batch_of_trials(
+                    ax_client=ax_client,
+                    parallelism=current_parallelism,
+                    num_trials_so_far=num_trials_so_far,
+                    num_max_trials_to_do=num_trials_left,
+                )
+                batch_of_trials_to_launch = batch_of_trials[:num_trials_left]
+
+                log.info(
+                    "AxSweeper is launching {} jobs".format(
+                        len(batch_of_trials_to_launch)
+                    )
+                )
+                overrides = [x.overrides for x in batch_of_trials_to_launch]
+
+                self.sweep_over_batches(
+                    ax_client=ax_client,
+                    overrides=overrides,
+                    batch_of_trials_to_launch=batch_of_trials_to_launch,
+                )
+
+                num_trials_so_far += len(batch_of_trials_to_launch)
+                num_trials_left -= len(batch_of_trials_to_launch)
+
+                best = ax_client.get_best_parameters()
+                metric = best[1][0][ax_client.objective_name]
+
+                if self.early_stopper.should_stop(metric, best[0]):
+                    num_trials_left = -1
+                    break
+
+            current_parallelism_index += 1
+
+        with open(f"{self.sweep_dir}/optimization_results.yaml", "w") as f:
+            results_to_serialize = {"optimizer": "ax", "ax": best}
+            results_to_serialize = json.loads(json.dumps(results_to_serialize))
+            # This step is to convert all the numpy floats into python floats
+            OmegaConf.save(OmegaConf.create(results_to_serialize), f)
```

Comment:
Much code, This is way too complicated.
try those two lines instead of the entire with block.
```python
result = OmegaConf.create({"optimizer": "ax", "ax": best.tolist()})
OmegaConf.save(result, f"{self.sweep_dir}/optimization_results.yaml")
```

Reply:
This fails as `AttributeError: 'tuple' object has no attribute 'tolist'` and `best` is a tuple.
`result = OmegaConf.create({"optimizer": "ax", "ax": best})` fails as 
`omegaconf.errors.UnsupportedValueType: key ax: tuple is not a supported type`. 

I think it is fine to load and unload a dict as json. I can get rid of line 83 and merge 84 with 85.

Reply:
Fixed in https://github.com/facebookresearch/hydra/pull/434/commits/3ea96d5c6bfb6bae377d35448e4c1d133a96d1ce